### ISSUE TITLE
[Bug]: apps.plugin doesn't group all subprocesses recursively

### ISSUE THREAD

**Issue by xd003** [netdata/netdata#18730]
### Bug description

The `apps.plugin` does not recursively group all subprocesses under the defined process. It only groups the immediate children of a process, while the subprocesses of these children remain ungrouped.

Consider the following example where I have the rule `test: '*bash start.sh*'` in my `apps_groups.conf`.

Let's assume the process tree is structured as:

- `bash start.sh`
  - `subprocess1`
    - `child_subprocess1` (bash)
      - `java_app`
  - `subprocess2`
  - `subprocess3`

Currently, only the immediate children processes (`subprocess1`, `subprocess2`, and `subprocess3`) are grouped under `test`. However, the child processes of `subprocess1` such as `child_subprocess1` and `java_app` are incorrectly grouped under `bash`, rather than being included in `test`.

### Expected behavior

The `apps.plugin` should recursively group all subprocesses, including child and grandchild processes, under the defined parent process. In the example above, `child_subprocess1` and `java_app` should be grouped under `test` along with `subprocess1`, `subprocess2`, and `subprocess3`.

### Installation method

docker

### System info

```shell
Linux netcup 6.1.0-25-arm64 #1 SMP Debian 6.1.106-3 (2024-08-26) aarch64 GNU/Linux
/etc/os-release:PRETTY_NAME="Debian GNU/Linux 12 (bookworm)"
/etc/os-release:NAME="Debian GNU/Linux"
/etc/os-release:VERSION_ID="12"
/etc/os-release:VERSION="12 (bookworm)"
/etc/os-release:VERSION_CODENAME=bookworm
/etc/os-release:ID=debian
```

### Netdata build info

```shell
Packaging:
    Netdata Version ____________________________________________ : v1.99.0-267-nightly
    Installation Type __________________________________________ : oci
    Package Architecture _______________________________________ : aarch64
    Package Distro _____________________________________________ : unknown
    Configure Options __________________________________________ : dummy-configure-command
Default Directories:
    User Configurations ________________________________________ : /etc/netdata
    Stock Configurations _______________________________________ : /usr/lib/netdata/conf.d
    Ephemeral Databases (metrics data, metadata) _______________ : /var/cache/netdata
    Permanent Databases ________________________________________ : /var/lib/netdata
    Plugins ____________________________________________________ : /usr/libexec/netdata/plugins.d
    Static Web Files ___________________________________________ : /usr/share/netdata/web
    Log Files __________________________________________________ : /var/log/netdata
    Lock Files _________________________________________________ : /var/lib/netdata/lock
    Home _______________________________________________________ : /var/lib/netdata
Operating System:
    Kernel _____________________________________________________ : Linux
    Kernel Version _____________________________________________ : 6.1.0-25-arm64
    Operating System ___________________________________________ : Debian GNU/Linux
    Operating System ID ________________________________________ : debian
    Operating System ID Like ___________________________________ : unknown
    Operating System Version ___________________________________ : 12 (bookworm)
    Operating System Version ID ________________________________ : 12
    Detection __________________________________________________ : /host/etc/os-release
Hardware:
    CPU Cores __________________________________________________ : 6
    CPU Frequency ______________________________________________ : unknown
    RAM Bytes __________________________________________________ : 8322441216
    Disk Capacity ______________________________________________ : 274877906944
    CPU Architecture ___________________________________________ : aarch64
    Virtualization Technology __________________________________ : unknown
    Virtualization Detection ___________________________________ : none
Container:
    Container __________________________________________________ : docker
    Container Detection ________________________________________ : dockerenv
    Container Orchestrator _____________________________________ : none
    Container Operating System _________________________________ : Debian GNU/Linux
    Container Operating System ID ______________________________ : debian
    Container Operating System ID Like _________________________ : unknown
    Container Operating System Version _________________________ : 12 (bookworm)
    Container Operating System Version ID ______________________ : 12
    Container Operating System Detection _______________________ : /etc/os-release
Features:
    Built For __________________________________________________ : Linux
    Netdata Cloud ______________________________________________ : YES
    Health (trigger alerts and send notifications) _____________ : YES
    Streaming (stream metrics to parent Netdata servers) _______ : YES
    Back-filling (of higher database tiers) ____________________ : YES
    Replication (fill the gaps of parent Netdata servers) ______ : YES
    Streaming and Replication Compression ______________________ : YES (zstd lz4 gzip)
    Contexts (index all active and archived metrics) ___________ : YES
    Tiering (multiple dbs with different metrics resolution) ___ : YES (5)
    Machine Learning ___________________________________________ : YES
Database Engines:
    dbengine (compression) _____________________________________ : YES (zstd lz4)
    alloc ______________________________________________________ : YES
    ram ________________________________________________________ : YES
    none _______________________________________________________ : YES
Connectivity Capabilities:
    ACLK (Agent-Cloud Link: MQTT over WebSockets over TLS) _____ : YES
    static (Netdata internal web server) _______________________ : YES
    h2o (web server) ___________________________________________ : YES
    WebRTC (experimental) ______________________________________ : NO
    Native HTTPS (TLS Support) _________________________________ : YES
    TLS Host Verification ______________________________________ : YES
Libraries:
    LZ4 (extremely fast lossless compression algorithm) ________ : YES
    ZSTD (fast, lossless compression algorithm) ________________ : YES
    zlib (lossless data-compression library) ___________________ : YES
    Brotli (generic-purpose lossless compression algorithm) ____ : NO
    protobuf (platform-neutral data serialization protocol) ____ : YES (system)
    OpenSSL (cryptography) _____________________________________ : YES
    libdatachannel (stand-alone WebRTC data channels) __________ : NO
    JSON-C (lightweight JSON manipulation) _____________________ : YES
    libcap (Linux capabilities system operations) ______________ : NO
    libcrypto (cryptographic functions) ________________________ : YES
    libyaml (library for parsing and emitting YAML) ____________ : YES
    libmnl (library for working with netfilter) ________________ : YES
Plugins:
    apps (monitor processes) ___________________________________ : YES
    cgroups (monitor containers and VMs) _______________________ : YES
    cgroup-network (associate interfaces to CGROUPS) ___________ : YES
    proc (monitor Linux systems) _______________________________ : YES
    tc (monitor Linux network QoS) _____________________________ : YES
    diskspace (monitor Linux mount points) _____________________ : YES
    freebsd (monitor FreeBSD systems) __________________________ : NO
    macos (monitor MacOS systems) ______________________________ : NO
    statsd (collect custom application metrics) ________________ : YES
    timex (check system clock synchronization) _________________ : YES
    idlejitter (check system latency and jitter) _______________ : YES
    bash (support shell data collection jobs - charts.d) _______ : YES
    debugfs (kernel debugging metrics) _________________________ : YES
    cups (monitor printers and print jobs) _____________________ : NO
    ebpf (monitor system calls) ________________________________ : NO
    freeipmi (monitor enterprise server H/W) ___________________ : YES
    nfacct (gather netfilter accounting) _______________________ : NO
    perf (collect kernel performance events) ___________________ : YES
    slabinfo (monitor kernel object caching) ___________________ : YES
    Xen ________________________________________________________ : NO
    Xen VBD Error Tracking _____________________________________ : NO
Exporters:
    AWS Kinesis ________________________________________________ : NO
    GCP PubSub _________________________________________________ : NO
    MongoDB ____________________________________________________ : YES
    Prometheus (OpenMetrics) Exporter __________________________ : YES
    Prometheus Remote Write ____________________________________ : YES
    Graphite ___________________________________________________ : YES
    Graphite HTTP / HTTPS ______________________________________ : YES
    JSON _______________________________________________________ : YES
    JSON HTTP / HTTPS __________________________________________ : YES
    OpenTSDB ___________________________________________________ : YES
    OpenTSDB HTTP / HTTPS ______________________________________ : YES
    All Metrics API ____________________________________________ : YES
    Shell (use metrics in shell scripts) _______________________ : YES
Debug/Developer Features:
    Trace All Netdata Allocations (with charts) ________________ : NO
    Developer Mode (more runtime checks, slower) _______________ : NO
```

**Comment by ilyam8** [comment]
@xd003 hi. Provide a script to reproduce the issue. And apps_groups.conf and expected result.

**Comment by xd003** [comment]
@ilyam8 Deploy the docker container as follows as it can't be deployed without adding some mandatory fields in the env file. Its a telegram bot

1) `sudo git clone https://github.com/anasty17/mirror-leech-telegram-bot /opt/mltb`
2) Now, create a new file `/opt/mltb/config.env` with the following content
```
BOT_TOKEN = ""                  
OWNER_ID = ""         
TELEGRAM_API = ""         
TELEGRAM_HASH = ""
JD_EMAIL = ""
JD_PASS = ""
```
- `BOT_TOKEN`: The Telegram Bot Token that you got from [@BotFather](https://t.me/BotFather). `Str`
- `OWNER_ID`: The Telegram User ID of the Owner of the bot. `Int`
- `TELEGRAM_API`: This is to authenticate your Telegram account for downloading Telegram files. You can get this
  from <https://my.telegram.org>. `Int`
- `TELEGRAM_HASH`: This is to authenticate your Telegram account for downloading Telegram files. You can get this
  from <https://my.telegram.org>. `Str`
- JD_EMAIL & JD_PASS must be obtained after signing up for an account [here](https://my.jdownloader.org/login.html#register)

3) Add the following to `docker-compose.yml` & run - `docker compose up -d mltb --build`
```
  mltb:
    build: /opt/mltb
    command: bash start.sh
    container_name: mediabot
    restart: unless-stopped
```
---
On netdata's end, have the `apps_groups.conf` with a single line - `mltb: '*bash start.sh*'`
Now, notice how all the processes from the above command are grouped in netdata. You will see that only processes which are direct children of `bash start.sh` are grouped as mltb

![image](https://github.com/user-attachments/assets/7220e852-0b31-4c93-8ea8-47aa17d7396f)

For instance in above process tree, the `PID - 454591 (sh)` & `454592 (java)` are categorized as `bash`

**Comment by ilyam8** [comment]
Fixed in #18734. Thanks for reporting, @xd003.

**Comment by xd003** [comment]
Can confirm that it's working fine in the latest nightly
